File: src/com/minecraft/moonlake/api/packet/listener/PacketListener.java
/*
 * Copyright (C) 2016 The MoonLake Authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


package com.minecraft.moonlake.api.packet.listener;

/**
 * <h1>PacketListener</h1>
 * 数据包监听器接口（详细doc待补充...）
 *
 * @version 1.0
 * @author Month_Light
 */
public interface PacketListener {

    /**
     * 处理此数据包通道的数据包发送
     *
     * @param receiver 接收者
     * @param packet 数据包
     * @param cancellable 阻止器
     * @return 数据包
     */
    Object onPacketSend(Object receiver, Object packet, Cancellable cancellable);

    /**
     * 处理此数据包通道的数据包接收
     *
     * @param sender 发送者
     * @param packet 数据包
     * @param cancellable 阻止器
     * @return 数据包
     */
    Object onPacketReceive(Object sender, Object packet, Cancellable cancellable);
}

File: API/src/main/kotlin/com/minecraft/moonlake/api/packet/PacketListener.kt
/*
 * Copyright (C) 2016-Present The MoonLake (mcmoonlake@hotmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.minecraft.moonlake.api.packet

import org.bukkit.plugin.Plugin

interface PacketListener {

    val plugin: Plugin

    val priority: PacketListenerPriority

    val sendingTypes: Set<Class<out PacketOut>>

    val receivingTypes: Set<Class<out PacketIn>>

    /**
     * Execute asynchronously on non-main threads when the server sends packets to the client.
     * If the [event] is cancelled, the client will not receive the packet.
     */
    fun onSending(event: PacketEvent)

    /**
     * Execute asynchronously on non-main threads when the client sends packets to the server.
     * If the [event] is cancelled, the server will not receive the packet.
     */
    fun onReceiving(event: PacketEvent)

    /**
     * If an exception is executed when [onSending] or [onReceiving], this method is called.
     * @param [ex] Exception Cause
     */
    fun handlerException(ex: Exception)
}
